Automattic/woocommerce-payments
# Add validation for Multi-Currency settings updates (and minor text fixes)

Fixes #6335 #6185

### Changes proposed in this Pull Request

**Add `InvalidCurrencyException` class**

* This was added to be able to be thrown whenever an invalid currency is used.

**Add `InvalidCurrencyCodeException` class**

* This was added to be able to be thrown whenever an invalid currency code is used.

**Updates to `MultiCurrency` class**

* Add `get_store_currencies` method to return the available, enabled, and default currencies for the store.
* Add `get_single_currency_settings` method to return the settings for a single currency.
* Add `update_single_currency_settings` to update the settings for a single currency.
* Update the `set_enabled_currencies` method to only allow available currencies to be set as enabled. This also makes it so improper data is not able to be stored in the option in the database. These checks include exiting if a string is passed instead of an array, or if anything other than an available currency code is in the passed array.

The additions above allow code to be removed from the `RestController` class and become available for any additional uses.

**Updates to `RestController` class**

* Update REST routes to include `args` that include the passed args along with their type, format, and if they are required. This helps with validation and sanitization.
* Moved some "working/lifting" code to new methods in the `MultiCurrency` class, as mentioned above. This keeps this method more on point and lean.
* Updated all methods to use `rest_ensure_response` to properly return REST responses.
* The `update_enabled_currencies` will check for exceptions and return a proper `WP_Error` if found.

**Updates and additions to unit tests**

* `MultiCurrency` unit tests were updated accordingly.
* Tests were added for the `RestController` class, and one was set to be skipped. This is due to in order to get `available_currencies` the account has to return as connected to WCPay via Jetpack. This is noted in the tests themselves.
  * I tried, for far too long, to try to get the test to work, but I don't believe it is possible. The class being tested doesn't have the `MultiCurrency` class/object passed to it, which means `MultiCurrency` and all objects that build it are not able to be mocked. Since there are no mocks involved, we are not able to make it seem like the test store is connected to WCPay.

**Other updates**

* Fix a couple minor text strings in the admin to solve/fix the issues mentioned above.

### Testing instructions





**Testing the update currencies modal and update currencies button:**

* Activate/include the below snippets on your site.
* Go to WooCommerce > Settings > Multi-Currency.
* Confirm the button to modify currencies reads _Add/remove currencies_ and has the proper "default" button styling.
* Click _Add/remove currencies_ so that the modal opens, you should see _AAA_ at the top of the list.
* Deactivate/remove the snippets on/from your site.
* Select _AAA_ in the modal and click _Update selected_.
* You should get a snackbar notice that says, "Error updating selected currencies"
* You should also get an error logged in your `woocommerce-payments-xxx` log like: "ERROR Invalid currency/currencies passed to set_enabled_currencies: AAA"

Snippets:
```
add_filter( 'wcpay_multi_currency_available_currencies', function( $currencies ) {
	$currencies[] = 'AAA';
	return $currencies;
} );

add_filter( 'woocommerce_currencies', function( $currencies ) {
	$currencies['AAA'] = 'AAA';
	return $currencies;
} );
```

**Testing currencies getting updated via UPE payment methods:**

* Apply patch below.
* Enable UPE.
* Enable P24, and then _Save changes_.
* There should be no errors logged in either the `debug` or `woocommerce-payments-xxx` log , this should silently fail.

Patch:
```diff
diff --git a/client/payment-methods-map.tsx b/client/payment-methods-map.tsx
index 89ef891bc..3b53df396 100644
--- a/client/payment-methods-map.tsx
+++ b/client/payment-methods-map.tsx
@@ -115,7 +115,7 @@ const PaymentMethodInformationObject: Record<
                        'woocommerce-payments'
                ),
                icon: iconComponent( P24Icon, 'Przelewy24 (P24)' ),
-               currencies: [ 'EUR', 'PLN' ],
+               currencies: [ 'EUR', 'PLN', 'AAA' ],
                stripe_key: 'p24_payments',
                allows_manual_capture: false,
        },
diff --git a/includes/payment-methods/class-p24-payment-method.php b/includes/payment-methods/class-p24-payment-method.php
index 73e388e1e..c962e14ca 100644
--- a/includes/payment-methods/class-p24-payment-method.php
+++ b/includes/payment-methods/class-p24-payment-method.php
@@ -26,7 +26,7 @@ class P24_Payment_Method extends UPE_Payment_Method {
                $this->stripe_id   = self::PAYMENT_METHOD_STRIPE_ID;
                $this->title       = 'Przelewy24 (P24)';
                $this->is_reusable = false;
-               $this->currencies  = [ 'EUR', 'PLN' ];
+               $this->currencies  = [ 'EUR', 'PLN', 'AAA' ];
                $this->icon_url    = plugins_url( 'assets/images/payment-methods/p24.svg', WCPAY_PLUGIN_FILE );
        }
 
```

**Testing text during MC onboarding is correct:**

* You need to update the option `wcpay_multi_currency_setup_completed` in the database to `no`: `wp option update wcpay_multi_currency_setup_completed no`
* Go to Payments > Overview, scroll down and look for the note with the title "Sell worldwide in multiple currencies"
* Click the _Set up now_ button.
* On the next page choose a currency and add it to get to step 2.
* You should now see what is in the below screenshot.

![Screenshot 2023-06-07 at 1 41 23 PM](https://github.com/Automattic/woocommerce-payments/assets/4634416/ef846f7e-50d8-4755-ba30-e8943335c71e)


-------------------


**Post merge**
